File: standards/error-envelope/rule.ts
import type { Finding, GateContext } from "../../src/standards/types.js";

/**
 * Reference pack 2 (build spec M4.6): one standard, two languages. This is the
 * TypeScript half; `rule.py` is the Python half and they share `standard.yaml`,
 * so the envelope's shape is configured once.
 *
 * Flags error payloads built as ad-hoc object literals instead of through the
 * shared envelope factory. Accepts a call to the factory, because that is the
 * whole point of having one.
 */

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((v): v is string => typeof v === "string");
}

function asString(value: unknown, fallback: string): string {
  return typeof value === "string" ? value : fallback;
}

const rule = (context: GateContext): Finding[] => {
  const parsed = context.ast;
  if (parsed === null) return [];

  const { ts, sourceFile } = parsed;
  const factory = asString(context.config["envelope_factory"], "errorResponse");
  const envelopeType = asString(context.config["envelope_type"], "ErrorResponse");
  const errorKeys = new Set(asStringArray(context.config["error_keys"]).map((k) => k.toLowerCase()));
  const senders = new Set(asStringArray(context.config["response_senders"]));

  const findings: Finding[] = [];
  const seen = new Set<number>();

  const report = (node: import("typescript").Node, what: string): void => {
    const line = parsed.lineOf(node.getStart(sourceFile));
    if (seen.has(line)) return;
    seen.add(line);
    findings.push({
      line,
      column: parsed.columnOf(node.getStart(sourceFile)),
      message: `${what} builds an error payload as a bare object literal`,
      fix: `return \`${factory}(...)\` so the response matches the shared \`${envelopeType}\` envelope`,
    });
  };

  /** Does this object literal look like an error payload? */
  const isErrorPayload = (node: import("typescript").ObjectLiteralExpression): boolean => {
    for (const prop of node.properties) {
      const name = prop.name;
      if (name === undefined) continue;
      const text = ts.isIdentifier(name) || ts.isStringLiteral(name) ? name.text.toLowerCase() : null;
      if (text !== null && errorKeys.has(text)) return true;
    }
    return false;
  };

  /**
   * An object built by spreading the factory's result is still enveloped:
   * `{ ...errorResponse(e), traceId }` is a deliberate extension, not a bypass.
   */
  const spreadsFactory = (node: import("typescript").ObjectLiteralExpression): boolean =>
    node.properties.some(
      (prop) =>
        ts.isSpreadAssignment(prop) &&
        ts.isCallExpression(prop.expression) &&
        prop.expression.expression.getText(sourceFile).includes(factory),
    );

  const check = (node: import("typescript").Node, what: string): void => {
    if (!ts.isObjectLiteralExpression(node)) return;
    if (!isErrorPayload(node)) return;
    if (spreadsFactory(node)) return;
    report(node, what);
  };

  const visit = (node: import("typescript").Node): void => {
    // `return { error: "..." }`
    if (ts.isReturnStatement(node) && node.expression !== undefined) {
      check(node.expression, "return statement");
    }

    // `res.json({ error })`, `res.status(500).send({ error })`, `JSONResponse({...})`
    if (ts.isCallExpression(node)) {
      const callee = node.expression;
      const name = ts.isPropertyAccessExpression(callee)
        ? callee.name.text
        : ts.isIdentifier(callee)
          ? callee.text
          : "";
      if (senders.has(name)) {
        const first = node.arguments[0];
        if (first !== undefined) check(first, `\`${name}()\``);
      }
    }

    node.forEachChild(visit);
  };

  visit(sourceFile);
  return findings;
};

export default rule;

File: standards/error-envelope/rule.py
"""Python half of the `error-envelope` standard.

Shares `standard.yaml` with `rule.ts`, so the envelope's shape is configured
once and both languages agree. Flags error payloads built as bare dict literals
instead of through the shared factory.
"""

from __future__ import annotations

import ast
from typing import Any

from keel_gates.model import Finding, GateContext


def _config_list(config: dict[str, Any], key: str) -> list[str]:
    value = config.get(key)
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, str)]


def _config_str(config: dict[str, Any], key: str, fallback: str) -> str:
    value = config.get(key)
    return value if isinstance(value, str) else fallback


def _dict_keys(node: ast.Dict) -> set[str]:
    keys: set[str] = set()
    for key in node.keys:
        if isinstance(key, ast.Constant) and isinstance(key.value, str):
            keys.add(key.value.lower())
    return keys


def _spreads_factory(node: ast.Dict, factory: str) -> bool:
    """``{**error_response(exc), "trace_id": t}`` is an extension, not a bypass."""
    for key, value in zip(node.keys, node.values, strict=True):
        if key is not None:
            continue
        # key is None for `**expr` entries.
        if isinstance(value, ast.Call) and _call_name(value) == factory:
            return True
    return False


def _call_name(node: ast.Call) -> str:
    func = node.func
    if isinstance(func, ast.Name):
        return func.id
    if isinstance(func, ast.Attribute):
        return func.attr
    return ""


def rule(context: GateContext) -> list[Finding]:
    try:
        tree = ast.parse(context.source)
    except SyntaxError:
        # Mid-edit files do not parse. No findings is the right answer.
        return []

    config = context.config
    factory = _config_str(config, "python_envelope_factory", "error_response")
    envelope_type = _config_str(config, "envelope_type", "ErrorResponse")
    error_keys = {k.lower() for k in _config_list(config, "error_keys")}
    senders = set(_config_list(config, "response_senders"))

    findings: list[Finding] = []
    seen: set[int] = set()

    def report(node: ast.AST, what: str) -> None:
        line = getattr(node, "lineno", 0)
        if line in seen or line == 0:
            return
        seen.add(line)
        findings.append(
            Finding(
                line=line,
                column=getattr(node, "col_offset", 0) + 1,
                message=f"{what} builds an error payload as a bare dict literal",
                fix=(
                    f"return `{factory}(...)` so the response matches the shared "
                    f"`{envelope_type}` envelope"
                ),
            )
        )

    def check(node: ast.AST, what: str) -> None:
        if not isinstance(node, ast.Dict):
            return
        if not (_dict_keys(node) & error_keys):
            return
        if _spreads_factory(node, factory):
            return
        report(node, what)

    for node in ast.walk(tree):
        if isinstance(node, ast.Return) and node.value is not None:
            check(node.value, "return statement")
        elif isinstance(node, ast.Call):
            name = _call_name(node)
            if name in senders and node.args:
                check(node.args[0], f"`{name}()`")

    return findings
